session/sql_store.go

```diff
@@ -0,0 +1,839 @@
+					return fmt.Errorf("unable to insert "+
+						"feature config: %w", err)
+				}
+			}
+		}
+
+		// Write privacy flags.
+		for _, flag := range sess.PrivacyFlags {
+			err := db.InsertSessionPrivacyFlag(
+				ctx, sqlc.InsertSessionPrivacyFlagParams{
+					SessionID: dbID,
+					Flag:      int32(flag),
+				},
+			)
+			if err != nil {
+				return fmt.Errorf("unable to insert privacy "+
+					"flag: %w", err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		mappedSQLErr := db.MapSQLError(err)
+		var uniqueConstraintErr *db.ErrSqlUniqueConstraintViolation
+		if errors.As(mappedSQLErr, &uniqueConstraintErr) {
+			// Add context to unique constraint errors.
+			return nil, fmt.Errorf("session violates unique "+
+				"constraint: %w", uniqueConstraintErr)
+		}
+
+		return nil, fmt.Errorf("unable to add session: %w", err)
+	}
+
+	return sess, nil
+}
+
+// ListSessionsByType returns all sessions currently known to the store that
+// have the given type.
+//
+// NOTE: this is part of the Store interface.
+func (s *SQLStore) ListSessionsByType(ctx context.Context, t Type) ([]*Session,
+	error) {
+
+	var (
+		readTxOpts = db.NewQueryReadTx()
+		sessions   []*Session
+	)
+	err := s.db.ExecTx(ctx, &readTxOpts, func(db SQLQueries) error {
+		dbSessions, err := db.ListSessionsByType(ctx, int16(t))
+		if err != nil {
+			return fmt.Errorf("could not list sessions: %w", err)
+		}
+
+		for _, dbSess := range dbSessions {
+			sess, err := unmarshalSession(ctx, db, dbSess)
+			if err != nil {
+				return fmt.Errorf("could not unmarshal "+
+					"session: %w", err)
+			}
+
+			sessions = append(sessions, sess)
+		}
+
+		return nil
+	})
+
+	return sessions, err
+}
+
+// ListSessionsByState returns all sessions currently known to the store that
+// are in the given state.
+//
+// NOTE: this is part of the Store interface.
+func (s *SQLStore) ListSessionsByState(ctx context.Context, state State) (
+	[]*Session, error) {
+
+	var (
+		readTxOpts = db.NewQueryReadTx()
+		sessions   []*Session
+	)
+	err := s.db.ExecTx(ctx, &readTxOpts, func(db SQLQueries) error {
+		dbSessions, err := db.ListSessionsByState(ctx, int16(state))
+		if err != nil {
+			return fmt.Errorf("could not list sessions: %w", err)
+		}
+
+		for _, dbSess := range dbSessions {
+			sess, err := unmarshalSession(ctx, db, dbSess)
+			if err != nil {
+				return fmt.Errorf("could not unmarshal "+
+					"session: %w", err)
+			}
+
+			sessions = append(sessions, sess)
+		}
+
+		return nil
+	})
+
+	return sessions, err
+}
+
+// ShiftState updates the state of the session with the given ID to the "dest"
+// state.
+//
+// NOTE: this is part of the Store interface.
+func (s *SQLStore) ShiftState(ctx context.Context, alias ID, dest State) error {
+	var writeTxOpts db.QueriesTxOptions
+	return s.db.ExecTx(ctx, &writeTxOpts, func(db SQLQueries) error {
+		dbSession, err := db.GetSessionByAlias(ctx, alias[:])
+		if err != nil {
+			return fmt.Errorf("unable to get session: %w", err)
+		}
```

Comment:
This function returns:
`ErrSessionNotFound` 

For the current kvdb logic when no session for the alias exists, so I think we'd want to mimic that rather then returning the sqlc `NoRows` error to outside the API?